src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -44,10 +44,12 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     double lat = convertToDouble(latString);
     double lng = convertToDouble(lngString);
     MapLocation userLocation = new MapLocation(lat, lng);
-    smallCityService.createUserServiceWithGeolocation(userLocation);
-    } catch(NullPointerException e) {
+    String product = request.getParameter("product");
+    smallCityService.createUserServiceWithGeolocation(userLocation, product);
+    } catch (NullPointerException e) {
       String zip = request.getParameter("zipCode");
-      smallCityService.createUserServiceWithZip(zip);
+      String product = request.getParameter("product");
```

Comment:
Odd. This is a repeat of line 47.

Reply:
It's a repeat because only one block of code will be executed. If there is no geolocation, the try block will fail before the product is grabbed.

Reply:
Right, but this is a code smell. This is retrying logic in the catch block that was already attempted in the try block - which defeats the purpose of the try/catch. Looks like the NPE is thrown from request.getParameter(). How about centering the try/catch only around those calls, something like the following:
```
public void doGet() {
    String latString;
    String product;
    ....
    
    try {
        latString = request.getParameter("lat");
        product = request.getParameter("product");
    } catch (NPE e) {
        Logger.warning(e.getMessage() + "more explanation");
    }

    SearchObject searchObject = new searchObject(product);
    smallCirtService.createUser(...);
    .....
}
```
An if statement might also be helpful.

Reply:
I restructured to a try/catch/finally to make sure that the right code is still being executed after the catch while repeating less code. Let me know if this solution still needs work!